In [117]:
import numpy as np
import statsmodels.api as sm

# Model_1

$$
y=\beta_{1} x_{1}+\varepsilon
$$

In [118]:
X = np.array([0,1,2,3,4])
Y = np.array([1,2,3,2,1])

In [119]:
model = sm.OLS(Y,X)
results = model.fit()
results.params
#print(results.summary())
#results.tvalues

array([0.6])

In [120]:
print(results.summary())

                                 OLS Regression Results                                
Dep. Variable:                      y   R-squared (uncentered):                   0.568
Model:                            OLS   Adj. R-squared (uncentered):              0.461
Method:                 Least Squares   F-statistic:                              5.268
Date:                Sun, 14 Mar 2021   Prob (F-statistic):                      0.0834
Time:                        14:00:18   Log-Likelihood:                         -8.3314
No. Observations:                   5   AIC:                                      18.66
Df Residuals:                       4   BIC:                                      18.27
Df Model:                           1                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------

c:\users\andy\appdata\local\programs\python\python37\lib\site-packages\statsmodels\stats\stattools.py:75: ValueWarning: omni_normtest is not valid with less than 8 observations; 5 samples were given.
  "samples were given." % int(n), ValueWarning)


In [121]:
#squared error loss
print("squared error loss =",sum((results.predict(X)-Y)**2)/len(X))
#absolute error loss
print("absolute error loss =",sum(abs(results.predict(X)-Y))/len(X))
#L1.5 loss
print("L1.5 loss =",sum(abs(results.predict(X)-Y)**(1.5))/len(X))

squared error loss = 1.64
absolute error loss = 1.1600000000000001
L1.5 loss = 1.3634801626671098


# Model_2

$$
y=\beta_{0}+\beta_{1} x_{1}+\varepsilon
$$

In [122]:
X = sm.add_constant(np.array([0,1,2,3,4]))
Y = np.array([1,2,3,2,1])

In [123]:
model = sm.OLS(Y,X)
results = model.fit()
results.params
#print(results.summary())
#results.tvalues

array([1.80000000e+00, 1.38777878e-16])

In [124]:
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                 -0.333
Method:                 Least Squares   F-statistic:                     0.000
Date:                Sun, 14 Mar 2021   Prob (F-statistic):               1.00
Time:                        14:00:19   Log-Likelihood:                -5.6451
No. Observations:                   5   AIC:                             15.29
Df Residuals:                       3   BIC:                             14.51
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          1.8000      0.748      2.405      0.0

c:\users\andy\appdata\local\programs\python\python37\lib\site-packages\statsmodels\stats\stattools.py:75: ValueWarning: omni_normtest is not valid with less than 8 observations; 5 samples were given.
  "samples were given." % int(n), ValueWarning)


In [125]:
#squared error loss
print("squared error loss =",sum((results.predict(X)-Y)**2)/len(X))
#absolute error loss
print("absolute error loss =",sum(abs(results.predict(X)-Y))/len(X))
#L1.5 loss
print("L1.5 loss =",sum(abs(results.predict(X)-Y)**(1.5))/len(X))

squared error loss = 0.56
absolute error loss = 0.64
L1.5 loss = 0.5849006163624495


# (a)

$$
\begin{array}{|c|c|c|}
\hline \text { Model } & \beta_{0} & \beta_{1} \\
\hline \text { Model }_{1} & 0 & 0.6\\
\hline \text { Model }_{2} & 1.8& 0\\
\hline
\end{array}
$$

# (b)

$$
\begin{array}{|c|c|c|c|}
\hline \text { Model } & \text { squared error loss } & \text { absolute error loss } & L_{1.5} \text { loss } \\
\hline \text { Model }_{1} & 1.64 & 1.16 & 1.36 \\
\hline \text { Model }_{2} & 0.56 & 0.64 & 0.58 \\
\hline
\end{array}
$$

# (c)

### Only doing simple linear regression seems not fit the label Y so well. We may need to involve some non-linear model.